### ISSUE TITLE
"Your Projects" cannot be reloaded

### ISSUE THREAD

**Issue by wjk** [DotPurpleApp/DotPurple-App#4]
**Describe the bug**  
There needs to be a way to force DotPurple to reread the solution file from disk without removing it from the project list and adding it again. DotPurple could do it automatically whenever the file is changed, or there can be a manual command.

Furthermore, trying to use DotPurple with a solution that contains a reference to a project that has since been removed from it (and deleted from disk as well) causes chaos, to the level where I do not know if DotPurple will work properly any longer or not. Quitting and reopening the application will not cause it to reread the solution file. I must remove the solution from the project list and add it again, which stomps all over my custom settings.

**Screenshots**

<img width="1012" alt="bug repro" src="https://user-images.githubusercontent.com/1915951/107575220-a966a700-6bbd-11eb-8b4d-3e8f783ae93f.png">

The selected project in the screenshot is the one that is causing the problem.

**Desktop**  
 - OS: macOS
 - Version 11.2.1

**Comment by Deadpikle** [comment]
@wjk Thanks for the bug report.

You're right, projects within a solution aren't being reloaded, which is an obvious problem. The bigger issue is that large solutions take a while to load (loading info on projects in a solution is rather slow), so we try to cache all the info we can.

At minimum, we need to add a refresh button. We need to look into file monitoring to force a reload when the SLN file changes. Also need to figure out why loading projects in a solution is slow once the number of projects grows. Watching a file doesn't seem terribly difficult in C# land, so I think we could handle this fairly easily. I hope.

Will try to get at least a stop-gap fix into the next version of the software.

**Comment by Deadpikle** [comment]
OK, so this was a little trickier than expected due to C#'s `FileSystemWatcher` not doing things on the UI thread, but we have some things working now!

* Files are monitored for renames and deletion
* Files are checked for existance upon loading the project screen the first time
* C#'s `FileSystemWatcher` doesn't monitor parent folders, so in order to catch a "deleted" file if the parent folder changes (or if the thing moves entirely), every 60 seconds we scan all projects to make sure they are all sitting on disk where we expect them to be. (We don't auto-close tabs, here, so as to not lose work, so keep open tabs after moving files around at your own risk. In theory, once you locate the file, everything will work fine, so no need to auto-close tabs)
![watching-file](https://user-images.githubusercontent.com/5092399/107880966-11fc9f00-6eb0-11eb-8fab-05bb8ba7e33b.gif)

* We also allow for rescanning solutions to view NEW projects added to the solution since the time it was added to DotPurple.
![rescan-sln](https://user-images.githubusercontent.com/5092399/107880967-132dcc00-6eb0-11eb-8bfc-d99623c93d77.gif)

The one major caveat right now would be that renaming a project on disk doesn't do anything to the actual solution or project files, so technically the file rename will be caught in DotPurple, but it won't adjust the SLN. We might add this as a feature later on.

**Comment by wjk** [comment]
Good work on the bugfix, and very fast, too! I very much approve of the changes. I do have one question though.

Does the last part of your previous post mean that if I move or rename a csproj file in DotPurple, the solution file and reality will become out-of-sync? This is not a deal-breaker by any means (next time I load the solution in VS, I just fix the problem by hand). You did excellent work.

**Comment by Deadpikle** [comment]
Well, DotPurple itself doesn't yet support renaming projects/solutions, but if you rename/move a project on disk (Finder on macOS, Windows Explorer on Windows, etc.), yes, technically the solution file will be out of sync in theory. However, if you rename the project in VS itself (Windows), VS should fix up the solution reference to the project that was renamed, I believe. I'd assume VS on Mac would do the same.

In theory, if we detected a rename, to "sync" the solution to the rename, we'd have to modify the solution file itself either using the `dotnet` CLI (remove -> add) or modifying the sln file itself. This need to modify the solution for the user is probably going to be pretty rare, though -- people aren't modifying their project locations and renaming things every day, so a little bit of extra manual labor isn't too bad, IMO. 

Of course, all of this relies on DotPurple being open on the projects page when the action occurs; if DotPurple isn't open, then, when you open DotPurple again, it'll just complain until you tell it where the files were moved to.

Regardless, we've got a note on this internally for if/when we get around to it.

Once I finish testing, here, I'll be rolling out the new version live, so stay tuned. :)
